### DISCUSSION TITLE
Support of multi-session discs

### DISCUSSION THREAD

**Discussion by ggiopp** [speed47/dvdisaster#147] [Q&A] [answered]
Hi all.

I understand that dvdisaster works by augmenting an image with error correction data, which is meant to be burned on the disc all at once.

I was wondering if this approach could be extended to work also on a multi-session disc, i.e., you write a first session made of the data and its error correction data, then later add further sessions of the same kind until the disc is full.

In this way, a single disc could be used as storage for an automatically backup system, which would, for example, add new data daily until the disc becomes full. 

Is this "theoretically" possible? 

Thanks,
Giovanni

**Comment by speed47** [comment]
Hello,

I think this could theoretically be possible, but that would be an entirely new dvdisaster codec to develop (RS04?), that would add this data after each session. Or maybe create an additional session for each session with only the correction data.

However, this is not an easy patch, requires some good amount of R&D, and clearly hundreds of working hours. This is out of scope of my maintainership role as I defined it for this project: ensuring dvdisaster never stops working on current OSes, possibly adding a few tiny features here and there, and providing bugfixes. I'm leaving this open just in case somebody has the time and knowledge to work on this :)

**Comment by e-c-d** [comment, marked as answer]
It is currently possible to use multisession in RS02 mode, but the recovery process is very awkward. The idea is to apply RS02 to every session ISO before writing it.

1. Generate an ISO as you would normally. For example, see the "Incremental backup with accumulated trees" section in the xorrisofs docs.
2. Add ECC data at the end: `dvdisaster -c -i example.iso -n 30% -mRS02`
3. Burn the image.

When reading it back:

1. Get the list of sessions and offsets ("LSN", Logical Sector Number with sector size = 2048): `cd-info -C /dev/sr0` or `xorriso -as cdrecord dev=/dev/sr0 -msinfo`. For example, let's say you get

    ```
      #: MSF       LSN    Type   Green? Copy?
      1: 00:02:00  000000 data   false  no   
      2: 10:14:27  045927 data   false  no   
      3: 12:40:02  056852 data   false  no   
    170: 14:26:02  064802 leadout (145 MB raw, 126 MB formatted)
    ```

    Let's say you want to repair session 2 (the one in the middle). This session starts at LSN=045927 and ends somewhere before LSN=056852.

5. Copy the data from that session only: `ddrescue -b 2048 --cpass=1 --skip-size=65536,1048576 -i $((045927 * 2048)) -s $(((056852-045927) * 2048)) -o 0 /dev/sr0 dump.iso mapfile.bin`, for example.
6. Fix it: `dvdisaster -f -i dump.iso`. You will probably need to add the `--truncate` argument.
7. Now you will probably have to do this for every session and then assemble them all back into a single big file with the correct absolute positions, and then you will be able to `losetup` it and mount it normally.

Best of luck.

**Comment by ggiopp** [comment]
Thanks @speed47, I understand. By the way, I have a lot of old discs protected with dvdisaster, I really appreciate your maintenance effort. 👍 

@e-c-d Your workaround looks interesting. The recovery procedure does not seem so difficult, after all. I'll keep it in mind and I'll may give it a spin. Thanks 😉
